### PR TITLE
NIFI-7225 FetchSFTP processor routing to not.found error given when P…

### PR DESCRIPTION
**Pull request by tlsmith109**
…rivate Key Path property is invalid

Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR 
The FetchSFTP processor utilized StandardValidator FileExistsValidator which allowed for directory names to be valid and not full path name to a file resulting in "not.found" errors being generated on PrivateKey path value. Corrected StandardValidators.FileExistsValidator to declare validation as "false" if path is to a directory. This change required modification of Junit tests that were using "./" as file path values, replacing values with valid "junk" file path.

 Fixes bug NIFI-7225._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on JDK 8?
- [x] Have you verified that the full build is successful on JDK 11?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [x] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [x] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-commons/nifi-utils/src/main/java/org/apache/nifi/processor/util/StandardValidators.java [hunk 1]
```diff
@@ -503,6 +503,7 @@ public ValidationResult validate(final String subject, final String input, final
     };
 
     public static final Validator FILE_EXISTS_VALIDATOR = new FileExistsValidator(true);
+    public static final Validator DIRECTORY_EXISTS_VALIDATOR = new DirectoryExistsValidator(true,false);
```
Comment (NissimShiman) [inline nifi-commons/nifi-utils/src/main/java/org/apache/nifi/processor/util/StandardValidators.java:507]: @tlsmith109 It is best we don't create a new variable for this.  Rather we should use createDirectorExistsValidator() as this is the way it is commonly done to enforce a directory existing.  See [1] and [2]

[1] https://github.com/apache/nifi/blob/main/nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/GetFile.java#L115
[2] https://github.com/apache/nifi/blob/main/nifi-nar-bundles/nifi-standard-bundle/nifi-standard-reporting-tasks/src/main/java/org/apache/nifi/controller/MonitorDiskUsage.java#L55

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/TailFile.java [hunk 2]
```diff
@@ -131,7 +131,7 @@
             .displayName("Base directory")
             .description("Base directory used to look for files to tail. This property is required when using Multifile mode.")
             .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
-            .addValidator(StandardValidators.FILE_EXISTS_VALIDATOR)
+            .addValidator(StandardValidators.DIRECTORY_EXISTS_VALIDATOR)
```
Comment (NissimShiman) [inline nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/TailFile.java:134]: @tlsmith109 A new ticket should be opened if we want to modify TailFile as well as the ticket is specific to FetchSFTP

That being said, it looks like there is code that already makes sure this is a directory [1].  Although, at first glance I am unsure why this wasn't handled here, as your code is suggesting. 

[1] https://github.com/apache/nifi/blob/main/nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/TailFile.java#L277

**Review by NissimShiman — COMMENTED** [review]
@tlsmith109  Thank you for working this ticket!

File: nifi-commons/nifi-utils/src/main/java/org/apache/nifi/processor/util/StandardValidators.java (hunk 1 above)
Comment (tlsmith109) [reply nifi-commons/nifi-utils/src/main/java/org/apache/nifi/processor/util/StandardValidators.java:507]: Will remove, I just added for consistency.

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/TailFile.java (hunk 2 above)
Comment (tlsmith109) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/TailFile.java:134]: Didn't realize I left this in here, will remove

File: nifi-commons/nifi-utils/src/main/java/org/apache/nifi/processor/util/StandardValidators.java [hunk 3]
```diff
@@ -848,8 +849,8 @@ public ValidationResult validate(final String subject, final String value, final
             }
 
             final File file = new File(substituted);
-            final boolean valid = file.exists();
-            final String explanation = valid ? null : "File " + file + " does not exist";
+            final boolean valid = file.exists() & !file.isDirectory();
+            final String explanation = valid ? null : "File " + file + " does not exist or is directory";
             return new ValidationResult.Builder().subject(subject).input(value).valid(valid).explanation(explanation).build();
         }
```
[the remark this answers is not in the document]
Comment (NissimShiman) [reply nifi-commons/nifi-utils/src/main/java/org/apache/nifi/processor/util/StandardValidators.java:855]: After discussions with @tlsmith109, I stand corrected.  This is very much relevant to this ticket.

My only question is why the apache nifi community hasn't noticed this sooner.  Maybe they have, but have left it on purpose to make sure everything works with backward compatibility?

@pvillard31 I noticed when you were working on TailFile in https://issues.apache.org/jira/browse/NIFI-1170 , when wishing to enforce that a filename is not a directory, you chose to check the validation in  customValidate() [1]  as opposed to modifying FileExistsValidator to not validate directories.

Was this by design to allow FileExistsValidator to validate directories?

[1] https://github.com/apache/nifi/blob/main/nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/TailFile.java#L288-L293

[the remark this answers is not in the document]
Comment (NissimShiman) [reply nifi-commons/nifi-utils/src/main/java/org/apache/nifi/processor/util/StandardValidators.java:855]: @tlsmith109 Let's just put the file checking in customValidate() to be consistent with how it done in TailFile.java

[the remark this answers is not in the document]
Comment (NissimShiman) [reply nifi-commons/nifi-utils/src/main/java/org/apache/nifi/processor/util/StandardValidators.java:855]: @tlsmith See comment from @MikeThomsen [1] " Any refactored version would have to not change the current default behavior. So, what I'd suggest would be a second constructor that allows you to set a boolean flag to put in a "file-only" mode."

I agree with this assessment and recommend this as well for this ticket. 

[1] https://mail-archives.apache.org/mod_mbox/nifi-dev/202011.mbox/browser.  
Subject: Re: FILE_EXISTS_VALIDATOR allows directories as well, is this a feature or a bug?

**Comment by NissimShiman** [comment]
Compiled and tested on java 8 and java 11.   Bug is fixed on both.  If a directory is given as a filename, this is noticed at configuration time with the standard yellow warning triangle as it should be.

Code is well structured and correct.  It is ready to be merged into main.

LGTM +1

**Comment by markap14** [comment]
Thanks for the update @tlsmith109 ! And thanks for the review @NissimShiman . All looks good to me as well. +1 will merge to main.

**Pull request merged**